subprojects/composite-builds/src/main/java/org/gradle/composite/internal/IncludedBuildArtifactBuilder.java

```diff
@@ -16,56 +16,40 @@
 
 package org.gradle.composite.internal;
 
-import com.google.common.collect.LinkedHashMultimap;
-import com.google.common.collect.Multimap;
 import org.gradle.api.GradleException;
 import org.gradle.api.artifacts.component.BuildIdentifier;
-import org.gradle.initialization.IncludedBuildExecuter;
+import org.gradle.initialization.IncludedBuildTaskGraph;
 import org.gradle.initialization.ReportedException;
 import org.gradle.internal.component.model.ComponentArtifactMetadata;
 import org.gradle.internal.exceptions.Contextual;
 import org.gradle.internal.exceptions.LocationAwareException;
 
-import java.util.Collection;
-
 class IncludedBuildArtifactBuilder {
-    private final Multimap<BuildIdentifier, String> tasksForBuild = LinkedHashMultimap.create();
-    private final IncludedBuildExecuter includedBuildExecuter;
-
-    IncludedBuildArtifactBuilder(IncludedBuildExecuter includedBuildExecuter) {
-        this.includedBuildExecuter = includedBuildExecuter;
-    }
+    private final IncludedBuildTaskGraph includedBuildTaskGraph;
 
-    public void willBuild(ComponentArtifactMetadata artifact) {
-        if (artifact instanceof CompositeProjectComponentArtifactMetadata) {
-            CompositeProjectComponentArtifactMetadata compositeBuildArtifact = (CompositeProjectComponentArtifactMetadata) artifact;
-            BuildIdentifier buildId = getBuildIdentifier(compositeBuildArtifact);
-            addTasksForBuild(buildId, compositeBuildArtifact);
-        }
+    IncludedBuildArtifactBuilder(IncludedBuildTaskGraph includedBuildTaskGraph) {
+        this.includedBuildTaskGraph = includedBuildTaskGraph;
     }
 
     public void build(BuildIdentifier requestingBuild, ComponentArtifactMetadata artifact) {
         if (artifact instanceof CompositeProjectComponentArtifactMetadata) {
             CompositeProjectComponentArtifactMetadata compositeBuildArtifact = (CompositeProjectComponentArtifactMetadata) artifact;
-            BuildIdentifier buildId = getBuildIdentifier(compositeBuildArtifact);
-            Collection<String> tasksToExecute = addTasksForBuild(buildId, compositeBuildArtifact);
-            execute(requestingBuild, buildId, tasksToExecute);
+            BuildIdentifier targetBuild = getBuildIdentifier(compositeBuildArtifact);
+            for (String taskName : compositeBuildArtifact.getTasks()) {
+                includedBuildTaskGraph.addTask(requestingBuild, targetBuild, taskName);
+            }
+            execute(targetBuild);
```

Comment:
I don't know if this is a big issue, but I think there's a chance for something weird like:

In parallel, two tasks need two separate project artifacts from the same included build, call them A and B.

We add the tasks for A's artifact and before we start executing the build, we also add B's tasks (from another thread).  The first included build execution will then build A and B.  The second one will block waiting on the first.  I think that means it would appear as if the first task caused by A and B to be built.

I looked at the way we lock things and I don't see anything obviously race-y that will break things.